### PR TITLE
[RelayerProxy] Test relayer proxy startup

### PR DESCRIPTION
**Pull request by red-0ne**
### Human Summary

Unit test the relayer proxy behavior

### Summary generated by Reviewpad on 07 Dec 23 00:50 UTC
This pull request includes the following changes:

1. The file `ring_signer.go` has changes related to the `Sign` function, including a parameter type change and the removal of a length check.
2. A new file `sessionquerier.go` has been added, which contains a struct and a method for querying session information.
3. The file `errors.go` has a new error `ErrQueryRetrieveSession` added.
4. The file `simple_signer.go` has changes to the `Sign` function signature, specifically the parameter type change.
5. The file `jsonrpc.go` has changes related to error handling in JSON RPC responses.
6. The file `relay_verifier.go` has changes in the `verifyResponse` function related to signable bytes and signature verification.
7. The file `cmd.go` has changes related to dependencies and the `setupRelayerDependencies` function.
8. The file `appquerier.go` has changes related to imports and function parameters.
9. The file `relay_signer.go` has changes in the `SignRelayResponse` function related to signable bytes.
10. The file `relay.go` has changes related to imports, signable bytes, and hashing.
11. The file `interface.go` has a change in the `Sign` method signature.
12. The file `accquerier.go` has changes related to the creation of a test account query client.
13. The file `relay_builders.go` has a new debugging log message added.
14. The file `supplierquerier.go` has changes related to dependencies and querying supplier information.
15. The file `interface.go` has changes related to client interfaces for querying on-chain information.
16. The file `sessionquerier.go` has a new function for creating a mock session query client.
17. The file `server_builder.go` has changes to the `BuildProvidedServices` function regarding supplier querying.

These changes involve various modifications to existing code, addition of new code, and enhancements to error handling, security, and query functionality.

## Issue

- #126 

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/relayer/cmd/cmd.go [hunk 1]
```diff
@@ -306,6 +311,24 @@ func newSupplyRelayerProxyFn(
...
+			return nil, err
+		}
+
+		clientCtx := cosmosclient.Context(queryClientCtx)
+
+		deps = depinject.Configs(deps, depinject.Supply(
+			accounttypes.NewQueryClient(clientCtx),
+			suppliertypes.NewQueryClient(clientCtx),
+			sessiontypes.NewQueryClient(clientCtx),
+			apptypes.NewQueryClient(clientCtx),
```
Comment (h5law) [inline pkg/relayer/cmd/cmd.go:328]: I think we should be using this pattern for these types as seen https://github.com/pokt-network/poktroll/pull/187/files#diff-74521651895cccd4d3e3a00792bc7eac38c0d396e86276d226a68c88381f1561:~:text=%40%40%20%2D0%2C0%20%2B1%2C57%20%40%40-,package%20query,-import%20( like we discussed off line.

And the depinject supplier functions like these https://github.com/pokt-network/poktroll/pull/187/files#diff-b0f30e8952edb9a9885d46fa8e6a353fe491fa4d55ddb9104539832599fcdca6:~:text=func%20NewSupplyAccountQuerierFn( which will go into the `cmd.go` supply config section

File: pkg/relayer/proxy/jsonrpc.go [hunk 2]
```diff
@@ -157,7 +157,7 @@ func (jsrv *jsonRPCServer) serveHTTP(ctx context.Context, request *http.Request)
 	relayHTTPRequest := &http.Request{
 		Method: request.Method,
 		Header: request.Header,
-		URL:    &jsrv.proxiedServiceEndpoint,
+		URL:    jsrv.proxiedServiceEndpoint,
```
Comment (h5law) [inline pkg/relayer/proxy/jsonrpc.go:160]: Make sure to merge with main now that #189 has merged this is changed in name
```suggestion
		URL:    sync.proxiedServiceEndpoint,
```

File: pkg/relayer/proxy/proxy.go [hunk 3]
```diff
@@ -108,20 +107,19 @@ func NewRelayerProxy(
 		deps,
 		&rp.clientCtx,
 		&rp.blockClient,
+		&rp.accountsQuerier,
+		&rp.supplierQuerier,
+		&rp.sessionQuerier,
+		&rp.applicationQuerier,
```
Comment (h5law) [inline pkg/relayer/proxy/proxy.go:113]: See the comment above (https://github.com/pokt-network/poktroll/pull/199/files#r1405127660) this should probably be interfaces that abstract the chain interaction.

See https://github.com/pokt-network/poktroll/pull/187/files#diff-93eab163375000803044e29070dbef1b1203622442eae9619001e19019ee1afe:~:text=type-,AccountQueryClient,-interface%20%7B for examples of the ones from #187

File: pkg/relayer/proxy/proxy.go [hunk 4]
```diff
@@ -189,7 +187,7 @@ func (rp *relayerProxy) validateConfig() error {
 		return ErrRelayerProxyUndefinedSigningKeyName
 	}
 
-	if rp.proxiedServicesEndpoints == nil {
+	if rp.proxiedServicesEndpoints == nil || len(rp.proxiedServicesEndpoints) == 0 {
```
Comment (h5law) [inline pkg/relayer/proxy/proxy.go:190]: Nice catch

File: pkg/relayer/proxy/proxy.go [hunk 5]
```diff
@@ -26,7 +25,7 @@ var _ relayer.RelayerProxy = (*relayerProxy)(nil)
 type (
 	serviceId            = string
 	relayServersMap      = map[serviceId][]relayer.RelayServer
-	servicesEndpointsMap = map[serviceId]*url.URL
+	ServicesEndpointsMap = map[serviceId]*url.URL
```
Comment (h5law) [inline pkg/relayer/proxy/proxy.go:28]: Why are we exporting this? Please leave a comment to explain <3

File: pkg/relayer/proxy/relay_builders.go [hunk 6]
```diff
@@ -19,9 +20,16 @@ func (jsrv *jsonRPCServer) newRelayRequest(request *http.Request) (*types.RelayR
 	log.Printf("DEBUG: Unmarshaling relay request...")
 	var relayReq types.RelayRequest
 	if err := relayReq.Unmarshal(requestBz); err != nil {
+		log.Printf("DEBUG: Unmarshaling relay request failed: %v", err)
 		return nil, err
 	}
 
+	if relayReq.Meta == nil {
```
Comment (h5law) [inline pkg/relayer/proxy/relay_builders.go:27]: :fire:

File: pkg/relayer/interface.go [hunk 7]
```diff
@@ -1,3 +1,8 @@
+//go:generate mockgen -destination=../../testutil/mockrelayer/account/query_client.go -package=mockaccount github.com/cosmos/cosmos-sdk/x/auth/types QueryClient
+//go:generate mockgen -destination=../../testutil/mockrelayer/application/query_client.go -package=mockapp github.com/pokt-network/poktroll/x/application/types QueryClient
+//go:generate mockgen -destination=../../testutil/mockrelayer/session/query_client.go -package=mocksession github.com/pokt-network/poktroll/x/session/types QueryClient
+//go:generate mockgen -destination=../../testutil/mockrelayer/supplier/query_client.go -package=mocksupplier github.com/pokt-network/poktroll/x/supplier/types QueryClient
+//go:generate mockgen -destination=../../testutil/mockrelayer/keyring/keyring.go -package=mockkeyring github.com/cosmos/cosmos-sdk/crypto/keyring Keyring
```
Comment (h5law) [inline pkg/relayer/interface.go:5]: These should become interfaces we define IMO #187 abstracting the cosmos-sdk stuff makes it simplier as discussed offline

File: testutil/mockrelayer/account/mocks.go [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package mockaccount
```
Comment (h5law) [inline testutil/mockrelayer/account/mocks.go:1]: Same as above comments I think adopting the pattern used in #187, we can avoid mocking cosmos-sdk components all together (we just need their interfaces)

File: testutil/testproxy/relayerproxy.go [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+	ApplicationPrivateKey *secp256k1.PrivKey
+}
+
+// ValidPayload is a helper variable for tests that need a valid payload
+var ValidPayload = &servicetypes.JSONRPCRequestPayload{
+	Method:  "someMethod",
+	Id:      1,
+	Jsonrpc: "2.0",
+	Params:  []string{"someParam"},
+}
```
Comment (h5law) [inline testutil/testproxy/relayerproxy.go:95]: Please merge with main
```suggestion
var ValidPayload = []byte(`{"method":"someMethod","id":1,"jsonrpc":"2.0","params":["someParam"]}`)
```
See https://github.com/pokt-network/poktroll/blob/main/pkg/partials/partials_test.go for similar stuff

**Review by h5law — DISMISSED** [review]
Screenshot 2023-11-25 at 15 01 44

Havent done a full review yet, just the two test files to go. Overall looks good, recommend merging with main as this changes how payloads are dealt with before addressing anything else. Left comments about what we spoke on earlier regarding the QueryClient stuff and wrapping them in interfaces to abstract the chain/cosmos logic.

File: pkg/relayer/proxy/proxy_test.go [hunk 10]
```diff
@@ -1,3 +1,424 @@
...
+	req.Meta.Signature = nil
+	return testproxy.MarshalAndSend(test, relayerProxyUrl, req)
+}
+
+func sendRequestWithInvalidRingSignature(
+	t *testing.T,
+	test *testproxy.TestBehavior,
+) (errorCode int32, errorMessage string) {
+	req := testproxy.GenerateRelayRequest(test, appPrivateKey, "service1", 1, validPayload)
+	req.Meta.Signature = []byte("invalid ring signature")
```
Comment (Olshansk) [inline pkg/relayer/proxy/proxy_test.go:334]: NIT: Remove `ring` from the function name, error message and test case. It should be agnostic to the type of signature under the hood.

File: pkg/relayer/proxy/proxy_test.go [hunk 11]
```diff
@@ -1,3 +1,424 @@
...
+}
+
+func sendRequestWithNonStakedApplicationAddress(
+	t *testing.T,
+	test *testproxy.TestBehavior,
+) (errCode int32, errorMessage string) {
+	randomPrivKey := secp256k1.GenPrivKey()
+	req := testproxy.GenerateRelayRequest(test, randomPrivKey, "service1", 1, validPayload)
+
+	// Have a legit signature from the non staked key
```
Comment (Olshansk) [inline pkg/relayer/proxy/proxy_test.go:363]:
```suggestion
	// Have a valid signature from the non staked key
```

File: testutil/testproxy/relayerproxy.go [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+
+			supplierAddress := supplierAccAddress.String()
+			sessionSuppliers = append(sessionSuppliers, supplierAddress)
+		}
+
+		testqueryclients.AddToExistingSessions(
+			test.t,
+			appAddress,
+			"service1",
+			1,
```
Comment (Olshansk) [inline testutil/testproxy/relayerproxy.go:208]: Move magic numbers to local constants.

File: testutil/testproxy/relayerproxy.go [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+	reader := io.NopCloser(bytes.NewReader(reqBz))
+	res, err := http.DefaultClient.Post(url, "application/json", reader)
+	require.NoError(test.t, err)
+	require.NotNil(test.t, res)
+
+	return GetRelayResponseError(test.t, res)
+}
+
+// GetRelayResponseError returns the error code and message from the relay response
+// if the response is not an error, it returns 0, ""
```
Comment (Olshansk) [inline testutil/testproxy/relayerproxy.go:232]:
```suggestion
// GetRelayResponseError returns the error code and message from the relay response.
// If the response is not an error, it returns `0, ""`.
```

File: testutil/testproxy/relayerproxy.go [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+	req *servicetypes.RelayRequest,
+	appPrivateKey *secp256k1.PrivKey,
+) []byte {
+	publicKey := appPrivateKey.PubKey()
+	curve := ring_secp256k1.NewCurve()
+
+	point, err := curve.DecodeToPoint(publicKey.Bytes())
+	require.NoError(t, err)
+
+	// At least two points are required to create a ring signer
```
Comment (Olshansk) [inline testutil/testproxy/relayerproxy.go:269]:
```suggestion
	// At least two points are required to create a ring signer
	// so we are reusing the same key for it
```

File: testutil/testproxy/relayerproxy.go [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+
+	scalar, err := curve.DecodeToScalar(appPrivateKey.Bytes())
+	require.NoError(t, err)
+
+	signer := signer.NewRingSigner(pointsRing, scalar)
+
+	signableBz, err := req.GetSignableBytes()
+	require.NoError(t, err)
+
+	hash := crypto.Sha256(signableBz)
```
Comment (Olshansk) [inline testutil/testproxy/relayerproxy.go:282]: We should be signing `signableBz`. Why is there an extra hash here?

File: testutil/testproxy/relayerproxy.go [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+
+	return signature
+}
+
+// GetAddressFromPrivateKey returns the address of the provided private key
+func GetAddressFromPrivateKey(test *TestBehavior, privKey *secp256k1.PrivKey) string {
+	applicationPublicKey, err := codectypes.NewAnyWithValue(privKey.PubKey())
+
+	require.NoError(test.t, err)
+	record := &keyringtypes.Record{Name: "app1", PubKey: applicationPublicKey}
```
Comment (Olshansk) [inline testutil/testproxy/relayerproxy.go:294]: There must be a way to get the address without creating a keyringRecord, no?

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/relayer/proxy/proxy_test.go [hunk 17]
```diff
@@ -1,3 +1,424 @@
...
+				proxy.WithProxiedServicesEndpoints(proxiedServices),
+			)
+			require.NoError(t, err)
+
+			go rp.Start(ctx)
+			time.Sleep(100 * time.Millisecond)
+
+			errCode, errMsg := tt.inputScenario(t, test)
+			require.Equal(t, tt.expectedErrCode, errCode)
+			require.True(t, strings.HasPrefix(errMsg, tt.expectedErrMsg))
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply pkg/relayer/proxy/proxy_test.go:286]: This is not a go error but a JsonRPC one, which is just a string

File: testutil/testproxy/relayerproxy.go (hunk 15 above)
Comment (red-0ne) [reply testutil/testproxy/relayerproxy.go:282]: `req.GetSignableBytes` just marshals the `RelayRequest` (excluding the signature field). Which means it results in an arbitrary length byte slice. We need to have a 32 bytes length content to sign.

File: pkg/relayer/proxy/proxy_test.go [hunk 18]
```diff
@@ -16,35 +16,59 @@ import (
...
+	validJsonRPCPayload []byte
 
-	supplierKeyName = "supplierKeyName"
-	appPrivateKey   = secp256k1.GenPrivKey()
-	// TODO_TECHDEBT: Source relayerProxyUrl from its config file once
+	// TODO_TECHDEBT(@okdas, @red-0ne): Source relayerProxyUrl from its config file once
 	// RelayerProxy is building its servers from the provided config file
+	relayerProxyUrl string
+
+	supplierKeyName   string
```
Comment (Olshansk) [inline pkg/relayer/proxy/proxy_test.go:32]:
```suggestion
	// helpers used for testing that are initialized in init
	supplierKeyName   string
```

File: pkg/relayer/proxy/proxy_test.go [hunk 19]
```diff
@@ -16,35 +16,59 @@ import (
...
+	supplierEndpoints []*sharedtypes.SupplierEndpoint
+	appPrivateKey     *secp256k1.PrivKey
+	proxiedServices   map[string]*url.URL
+
+	defaultRelayerProxyBehavior []func(*testproxy.TestBehavior)
+)
+
+func init() {
+	validJsonRPCPayload = []byte(`{"method":"someMethod","id":1,"jsonrpc":"2.0","params":["someParam"]}`)
```
Comment (Olshansk) [inline pkg/relayer/proxy/proxy_test.go:41]:
```suggestion
	// TODO_IMPROVE: Add support for oher types of requests as well
	validJsonRPCPayload = []byte(`{"method":"someMethod","id":1,"jsonrpc":"2.0","params":["someParam"]}`)
```

File: pkg/relayer/proxy/proxy_test.go [hunk 20]
```diff
@@ -16,35 +16,59 @@ import (
...
+	supplierKeyName = "supplierKeyName"
+	supplierEndpoints = []*sharedtypes.SupplierEndpoint{
+		{
+			Url: "http://",
```
Comment (Olshansk) [inline pkg/relayer/proxy/proxy_test.go:45]: Confirming that this is intended to be empty?

File: pkg/relayer/proxy/proxy_test.go [hunk 21]
```diff
@@ -112,19 +137,19 @@ func TestRelayerProxy_NoProxiedServices(t *testing.T) {
 }
 
 // RelayerProxy should fail to start if it cannot spawn a server for the
-// services it advertized on-chain
-// TODO_TECHDEBT: Re-enable this test once the RelayerProxy is building its
-// servers from the provided config file
+// services it advertized on-chain.
+// TODO_TECHDEBT(@okdas, @red-0ne): Re-enable this test once the RelayerProxy
+// is building its servers from the provided config file.
 //func TestRelayerProxy_UnsupportedRpcType(t *testing.T) {
```
Comment (Olshansk) [inline pkg/relayer/proxy/proxy_test.go:143]: Can you do the following.

1. Uncomment the test
2. Use `t.skip("")` in the very beginning of the test so it's not executed
3. Add the TODO comment inside the message inside t.skip

File: testutil/testproxy/jsonrpc.go [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+// JSONRpcErrorReply is the error reply struct for the JSON RPC response
+type JSONRpcErrorReply struct {
+	Id      int32  `json:"id"`
+	Jsonrpc string `json:"jsonrpc"`
+	Error   *JSONRpcError
+}
+
+func prepareJsonRPCPayload(serviceId string) string {
```
Comment (Olshansk) [inline testutil/testproxy/jsonrpc.go:18]:
```suggestion
// prepareJsonRPCPayload prepares a hard-coded JsonPRC payload for a specific request.
func prepareJsonRPCPayload(serviceId string) string {
```

File: testutil/testproxy/jsonrpc.go [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+	return fmt.Sprintf(`{"jsonrpc":"2.0","id":1,"result":"%s"}`, serviceId)
```
Comment (Olshansk) [inline testutil/testproxy/jsonrpc.go:19]: Don't we want the hardcoded request to be a proper string? A bit confused by this.

Can you explain, #PUC or change the code?

File: x/session/keeper/session_hydrator.go [hunk 24]
```diff
@@ -237,3 +237,15 @@ func sha3Hash(bz []byte) []byte {
 	hasher.Write(bz)
 	return hasher.Sum(nil)
 }
+
+func SessionIdBzToString(appPubKey, serviceId, blockHash string, blockHeight int64) string {
```
Comment (Olshansk) [inline x/session/keeper/session_hydrator.go:241]: 1. Godoc
2. Make sure to re-use in the actual on-chain code as well so there's only 1 source of truth for the implementaiton

File: x/session/keeper/session_hydrator.go [hunk 25]
```diff
@@ -237,3 +237,15 @@ func sha3Hash(bz []byte) []byte {
...
+	appPubKeyBz := []byte(appPubKey)
+	serviceIdBz := []byte(serviceId)
+	blockHashBz := []byte(blockHash)
+	sessionHeightBz := make([]byte, 8)
```
Comment (Olshansk) [inline x/session/keeper/session_hydrator.go:245]:
```suggestion

	sessionHeightBz := make([]byte, 8)
```

File: testutil/testproxy/relayerproxy.go [hunk 26]
```diff
@@ -150,13 +138,20 @@ func WithDefaultActors(
...
+			test.t,
+			supplierAddress,
+			"service1",
+			supplierEndpoints,
+		)
+	}
+}
+
+// WithDefaultApplication creates the default application actor for the test
+// and mock it is staked on-chain.
```
Comment (Olshansk) [inline testutil/testproxy/relayerproxy.go:152]: `and mock it is staked on-chain.`

Unclear what you intended to say. #PUC

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just a few more small comments and then we're good to got! Great work!

Make sure to also to also include the `e2e-tag` after you push your last set of changes.

File: pkg/relayer/proxy/proxy_test.go (hunk 20 above)
Comment (red-0ne) [reply pkg/relayer/proxy/proxy_test.go:45]: Yes, but will fill it to avoid confusion

File: pkg/relayer/proxy/proxy_test.go (hunk 21 above)
Comment (red-0ne) [reply pkg/relayer/proxy/proxy_test.go:143]: The test no longer depends on that TECHDEBT, will renable it for now.

File: pkg/relayer/proxy/proxy_test.go (hunk 19 above)
Comment (red-0ne) [reply pkg/relayer/proxy/proxy_test.go:41]: Moved to `testutil/testproxy/jsonrpc.go` under `PrepareJsonRPCRequestPayload`

File: x/session/keeper/session_hydrator.go (hunk 24 above)
Comment (red-0ne) [reply x/session/keeper/session_hydrator.go:243]: Please review its on-chain usage

File: testutil/testproxy/relayerproxy.go (hunk 15 above)
Comment (Olshansk) [reply testutil/testproxy/relayerproxy.go:274]: That's a great point but it means we have a large oversight in a fundamental component of our codebase. `GetSignableBytes` is going to be very important in:
- Signing relays
- Verifying relays
- Claim & Proof lifecycle
- Indexers
- Challenges
- Etc...

The fact that this logic (which is key to state transitions) is not centralized is a big problem IMO.

I think we need to create a new PR ASAP (and merge it in before this one) that updates `GetSignableBytes` to return a constant length string.

1. Wdyt?
2. Would you be able to do that?

cc @h5law

File: pkg/relayer/proxy/proxy_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply pkg/relayer/proxy/proxy_test.go:323]: Got it 👍

**Review by Olshansk — APPROVED** [review]
@red-0ne LGTM!

We need to run `devnet-e2e` and merge with main (the signable bytes change you introduced), but lgtm otherwise!

**Pull request merged**